storage/innobase/mem/mem0mem.c

```diff
@@ -404,6 +404,8 @@ mem_heap_create_block(
 		heap->total_size += len;
 	}
 
+	UNIV_MEM_FREE(block + 1, len - MEM_BLOCK_HEADER_SIZE);
```

Comment:
I think that this deserves a comment and an assertion. If I understood the intention correctly, I believe that we would want something like this:
```c
compile_time_assert(MEM_BLOCK_HEADER_SIZE >= sizeof *block);
UNIV_MEM_FREE(block + 1, len - sizeof *block);
```
Note: the `>=`  above is because of the definition:
```c
#define MEM_BLOCK_HEADER_SIZE	ut_calc_align(sizeof(mem_block_info_t),\
							UNIV_MEM_ALIGNMENT)
```

Reply:
Fixed.

Maybe it would be easier to add `MY_ALIGNED` and/or `static_assert` on `sizeof mem_block_info_t` and remove `MEM_BLOCK_HEADER_SIZE`?

Static assert like this can be used in a global scope too:
```c++
#define STATIC_ASSERT( condition, name )\
    typedef char assert_failed_ ## name [ (condition) ? 1 : -1 ];
```

Also, in future red zoned between individual allocation chunks would be useful too. They will catch overflows and underflows for every allocated chunk.